diozero-core/src/main/java/com/diozero/devices/AHT21.java

```diff
@@ -0,0 +1,204 @@
+            SleepUtil.sleepMillis(10);
+        }
+
+        /*
+        After receiving six bytes, the next byte is CRC check data, which the user can read as needed. If the receiver
+        needs CRC check, it will send an ACK reply after receiving the sixth byte, otherwise it will send a NACK reply.
+         */
+
+        i2c.writeByte(READ_COMMAND);
+        // 1st status byte,
+        // then 5 bytes with hum & temp,
+        // last byte crc (optional)
+        final byte[] resp = new byte[withCrc ? 7 : 6];
+        i2c.readBytes(resp);
+
+        return resp;
+    }
+
+    private float[] parseResponse(byte[] resp) {
+        if (resp.length > 6) { // check last byte CRC
+            final byte[] data = new byte[6];
+            System.arraycopy(resp, 0, data, 0, data.length);
+
+            final int computedCrc = computeCrc(data);
+            final int expectedCrc = toUnsignedInt(resp[6]);
+
+            if (computedCrc != expectedCrc) {
+                Logger.debug("CRC check failed");
+                return new float[]{-1F, -1F};
+            }
+        }
+
+        // skip status
+        // humidity is next with 20 bits
+        final int hum = (toUnsignedInt(resp[1]) << 16 | toUnsignedInt(resp[2]) << 8 | toUnsignedInt(resp[3])) >> 4;
+        // temperature is next with 20 bits
+        final int temp = (toUnsignedInt(resp[3]) & 0x0F) << 16 | toUnsignedInt(resp[4]) << 8 | toUnsignedInt(resp[5]);
+
+        // magic formula from datasheet
+        final float relativeHum = hum * 100F / 1024 / 1024;
+        final float temperature = temp * 200F / 1024 / 1024 - 50;
+
+        return new float[]{relativeHum, temperature};
+    }
+
+    /**
+     * Compute CRC for given data.
+     * The initial value of CRC is 0xFF.
+     * CRC check type: CRC8/MAXIM
+     * Polynomial: X8+X5+X4+1 Poly：0011 0001 0x31 When placed high, it becomes 1000 1100 0x8c
+     * Adapted from
+     * <a href="http://aosong.com/userfiles/files/software/AHT20-21%20DEMO%20V1_3(1).rar">datasheet reference</a>
+     *
+     * @param data the data
+     * @return the CRC
+     */
+    private int computeCrc(byte[] data) {
+        int crc = 0xFF;
+
+        for (byte current : data) {
+            int byteData = Byte.toUnsignedInt(current);
+            crc ^= byteData;
+            for (int index = 8; index > 0; --index) {
+                if ((crc & 0x80) == 0x80) {
+                    crc = crc << 1 ^ 0x31;
+                } else {
+                    crc = crc << 1;
+                }
+            }
+        }
+
+        return crc & 0xFF; // CRC is 1 byte
+    }
+
+    /**
+     * Read both humidity and temperature.
+     *
+     * @param verifyCrc whether to check CRC byte. If CRC fails then returned value is [-1F, -1F], but if it passes
+     *                  it increases measurement confidence
+     * @return [relative humidity, temperature (Celsius)]
+     */
+    public float[] getValues(boolean verifyCrc) {
```

Comment:
This appears to be a lot of nested calls that are effectively identical. For example, the `verifyCrc` parameter is _always_ `true`, so it's basically not necessary.

Reply:
Thank you for the feedback.
When I implemented I considered the existing BME280 code that you already have, which has a `float[] getValues()`, `float getTemperature()` and `float getRelativeHumidty()`. The last 2 are required because of the interfaces `ThermometerInterface` and `HygrometerInterface` but I also left the `getValues` public because it can read both at the same time. So that's why there are 3 get* methods and why the they return floats.

The `parseResponse` does indeed always wrap the `readMeasurement` which I find more readable/logically split but if you think it's better in a single method, I can modify that.

For CRC, I left a flag there because while I was going through  the datasheet and sample programs, I found that AHT21 has a lot in common with AHT20, but AHT20 does not support CRC ( but I don't have a AHT20 sensor). 
But, I agree with you, I guess you would always want to verify the CRC, if available. I am not sure, now that I think about it, what should be done if the check fails, right now it returns [-1, -1], but it might be better to throw an exception since -1 is a valid temperature and the user might use the sensor through the `ThermometerInterface`. What do you think?

Reply:
Hmmm - the issue I have with following the interfaces here is it requires **two** calls to the sensor. However, trying to cache a read value is not practical, so I withdraw my suggestion about the `get`.

The CRC probably should throw an exception, since "magic error return values" (as you note) aren't necessarily applicable.